File: test/vti/dashboard/src/main/webapp/js/time.js
/**
 * Copyright (c) 2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you
 * may not use this file except in compliance with the License. You may
 * obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

(function (moment) {

  /**
   * Renders a timestamp in the user timezone.
   * @param timestamp The long timestamp to render (in microseconds).
   * @param showTimezone True if the timezone should be rendered, false otherwise.
   * @returns the string-formatted version of the provided timestamp.
   */
  moment.prototype.renderTime = function (timestamp, showTimezone) {
    var time = moment(timestamp / 1000);
    var format = 'YYYY-M-DD H:mm:ss';
    if (!!showTimezone) {
        format = format + 'ZZ';
    }
    return time.format(format);
  }

  /**
   * Renders a date in the user timezone.
   * @param timestamp The long timestamp to render (in microseconds).
   * @param showTimezone True if the timezone should be rendered, false otherwise.
   * @returns the string-formatted version of the provided timestamp.
   */
  moment.prototype.renderDate = function (timestamp, showTimezone) {
    var time = moment(timestamp / 1000);
    var format = 'YYYY-M-DD';
    if (!!showTimezone) {
        format = format + 'ZZ';
    }
    return time.format(format);
  }

  /**
   * Renders a duration in the user timezone.
   * @param durationTimestamp The long duration to render (in microseconds).
   * @returns the string-formatted duration of the provided duration timestamp.
   */
  moment.prototype.renderDuration = function (durationTimestamp) {
    var fmt = 's[s]';
    var duration = moment.utc(durationTimestamp / 1000);
    if (duration.hours() > 0) {
      fmt = 'H[h], m[m], ' + fmt;
    } else if (duration.minutes() > 0) {
      fmt = 'm[m], ' + fmt;
    }
    return duration.format(fmt);
  }

})(moment);

File: system/bt/osi/src/time.cc
/******************************************************************************
 *
 *  Copyright 2015 Google, Inc.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at:
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 ******************************************************************************/

#define LOG_TAG "bt_osi_time"

#include <sys/time.h>
#include <time.h>

#include "osi/include/time.h"

uint32_t time_get_os_boottime_ms(void) {
  return (uint32_t)(time_get_os_boottime_us() / 1000);
}

uint64_t time_get_os_boottime_us(void) {
  struct timespec ts_now;
  clock_gettime(CLOCK_BOOTTIME, &ts_now);

  return ((uint64_t)ts_now.tv_sec * 1000000L) +
         ((uint64_t)ts_now.tv_nsec / 1000);
}

uint64_t time_gettimeofday_us(void) {
  struct timeval tv;
  gettimeofday(&tv, NULL);
  return static_cast<uint64_t>(tv.tv_sec) * 1000000ULL +
         static_cast<uint64_t>(tv.tv_usec);
}
